### PR TITLE
Add DRDA Protocol Decoder Support

### PR DESCRIPTION
**Pull request by AndrewDi**
Add DRDA Protocol Decoder Support

### CODE REVIEW

File: PacketDotNet/DrdaCodepointType.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
+﻿using System;
```
Comment (chmorgan) [inline PacketDotNet/DrdaCodepointType.cs]: Please include the licence header here and your information for the copyright statement.

File: .vs/VSWorkspaceState.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (chmorgan) [inline .vs/VSWorkspaceState.json]: Saw that you removed this file in a later commit, please squash those commits with this one so in the git history the files were never added in the first place.

File: PacketDotNet/DrdaDDMFields.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+PacketDotNet is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with PacketDotNet.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/*
+ *  Copyright 2009 Chris Morgan <chmorgan@gmail.com>
```
Comment (chmorgan) [inline PacketDotNet/DrdaDDMFields.cs]: Please update with your information as you authored this file

File: PacketDotNet/DrdaDDMPacket.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+PacketDotNet is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with PacketDotNet.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/*
+ * Copyright 2010 Chris Morgan <chmorgan@gmail.com>
```
Comment (chmorgan) [inline PacketDotNet/DrdaDDMPacket.cs]: And here.

File: PacketDotNet/DrdaDDMPacket.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                        var parameter = new DrdaDDMParameter()
+                        {
+                            Length = length,
+                            DrdaCodepoint = (DrdaCodepointType)BigEndianBitConverter.Big.ToUInt16(header.Bytes, offset + DrdaDDMFields.ParameterLengthLength)
+                        };
+                        
+                        if(parameter .DrdaCodepoint== DrdaCodepointType.DATA|| parameter.DrdaCodepoint == DrdaCodepointType.QRYDTA)
+                        {
+                            var startIndex = offset + DrdaDDMFields.ParameterLengthLength + DrdaDDMFields.ParameterCodePointLength;
+                            var strLength = length - 4;
```
Comment (chmorgan) [inline PacketDotNet/DrdaDDMPacket.cs]: Can you create a variable to replace this magic '4' with something that explains why its value is 4?

File: PacketDotNet/DrdaDDMPacket.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+                            parameter.Data = ASCIIEncoding.UTF8.GetString(header.Bytes, startIndex,strLength);
+                        }
+                        else
+                        {
+                            parameter.Data = StringConverter.EbcdicToAscii(header.Bytes, offset + DrdaDDMFields.ParameterLengthLength + DrdaDDMFields.ParameterCodePointLength, length - 2 - 2);
```
Comment (chmorgan) [inline PacketDotNet/DrdaDDMPacket.cs]: And here with the -2 and -2. I presume those are accounting for something in the data fields so two variables, one for each 2, would help describe why the value is 2.

File: PacketDotNet/DrdaPacket.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+PacketDotNet is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with PacketDotNet.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/*
+ * Copyright 2010 Chris Morgan <chmorgan@gmail.com>
```
Comment (chmorgan) [inline PacketDotNet/DrdaPacket.cs]: Your copyright here.

File: PacketDotNet/MiscUtil/Conversion/StringConverter.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
```
Comment (chmorgan) [inline PacketDotNet/MiscUtil/Conversion/StringConverter.cs]: license and copyright here.

File: PacketDotNet/TcpFields.cs [hunk 9]
```diff
@@ -22,8 +22,8 @@
 
 namespace PacketDotNet
 {
-    /// <summary> IP protocol field encoding information.
-    ///
+    /// <summary> 
+    /// IP protocol field encoding information.
```
Comment (chmorgan) [inline PacketDotNet/TcpFields.cs]: lets drop this change from this commit, it should go into a separate commit as it isn't related to the packet dissection you are adding.

File: PacketDotNet/TcpPacket.cs [hunk 10]
```diff
@@ -421,10 +421,29 @@ public TcpPacket(ByteArraySegment bas)
                 // the length of the payload is the total payload length
                 // above, minus the length of the tcp header
                 payloadPacketOrData.TheByteArraySegment.Length = newTcpPayloadLength;
+                this.DecodePayload();
             }
         }
 
         /// <summary>
+        /// Decode Payload to Support Drda procotol
```
Comment (chmorgan) [inline PacketDotNet/TcpPacket.cs]: Can we move the changes to add decoding to TcpPacket into another commit so we can discuss them separately?

File: PacketDotNet/TcpPacket.cs [hunk 11]
```diff
@@ -421,10 +421,29 @@ public TcpPacket(ByteArraySegment bas)
...
+        /// </summary>
+        /// <returns></returns>
+        public TcpPacket DecodePayload()
+        {
+            if(PayloadData==null || PayloadData.Length == 0)
+            {
+                return this;
+            }
+            if(PayloadData.Length>=10&&PayloadData[2]== 0xd0)
```
Comment (chmorgan) [inline PacketDotNet/TcpPacket.cs]: Can you modify this to make it more obvious what is going on? For instance is 10 the minimum length of a DRDA packet that you could get from a constant in DRDAFields? I'd say the same would help clarify what 0xd0 is as well.

File: Test/PacketType/DrdaPacketTest.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+PacketDotNet is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with PacketDotNet.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/*
+ *  Copyright 2010 Chris Morgan <chmorgan@gmail.com>
```
Comment (chmorgan) [inline Test/PacketType/DrdaPacketTest.cs]: Update copyright to you.

File: Test/PacketType/DrdaPacketTest.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                        Console.WriteLine(ddm.ToString(StringOutputType.Colored));
+                    }
+                }
+                packetIndex++;
+            }
+            dev.Close();
+
+        }
+
+        [Test]
```
Comment (chmorgan) [inline Test/PacketType/DrdaPacketTest.cs]: This should go into a separate test file as it is specifically testing the conversion you added. The other point I'd make is that you are duplicating the array of values here. Imo you could drop that as its just a copy of what you have in the conversion routine and simply include the input array of values and what you expect as the output. Duplicating the array here isn't buying us much, we already have that and can use that out of the other class you added without duplicating it here.

File: Test/PacketType/DrdaPacketTest.cs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+using PacketDotNet;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+
+namespace Test.PacketType
+{
+    class DrdaPacketTest
+    {
+        [Test]
+        public void TestSelectPacket()
```
Comment (chmorgan) [inline Test/PacketType/DrdaPacketTest.cs]: Excellent that you've added a test here. Can you compare some fields of the decoded packets against what you'd expect them to be? Right now you are just printing the data out so if the decoding was broken and printing out incorrect values the test would still pass. I typically use wireshark to retrieve values for the unit tests here as wireshark has a ton more development time put into it than these decoders have had.

File: Test/Test.csproj [hunk 15]
```diff
@@ -117,5 +118,7 @@
   <ItemGroup>
     <None Include="log4net.config" />
   </ItemGroup>
-  <ItemGroup />
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
```
Comment (chmorgan) [inline Test/Test.csproj]: Not sure what this service is but if its necessary then it can stay here.

File: PacketDotNet/DrdaDDMPacket.cs [hunk 16]
```diff
@@ -198,18 +206,20 @@ public override string ToString(StringOutputType outputFormat)
...
                     colorEscape,
                     paramter.Length,
                     paramter.DrdaCodepoint,
                     paramter.Data);
                 }
+                buffer.Append("}");
             }
             // append the base class output
-            buffer.Append(base.ToString(outputFormat));
+            //buffer.Append(base.ToString(outputFormat));
```
Comment (chmorgan) [inline PacketDotNet/DrdaDDMPacket.cs]: I prefer that we don't include commented out code. If you'd like to keep it you should uncomment it, if it can go you can just delete it.

File: PacketDotNet/MiscUtil/Conversion/StringConverter.cs [hunk 17]
```diff
@@ -35,6 +35,8 @@ public static string EbcdicToAscii(byte[] data,int offset,int length)
             {
                 try
                 {
+                    if (data[i] == 0x00)
```
Comment (chmorgan) [inline PacketDotNet/MiscUtil/Conversion/StringConverter.cs]: why skipping over zero/null values? Can you explain why you are continuing here but still processing the next byte? I'm not familiar with ebcdic.

File: PacketDotNet/DrdaDDMPacket.cs [hunk 18]
```diff
@@ -110,37 +110,45 @@ public DrdaCodepointType CodePoint
         }
 
         private List<DrdaDDMParameter> paramters;
+
```
Comment (chmorgan) [inline PacketDotNet/DrdaDDMPacket.cs]: You can fold these improvements into your initial commit if they are improving that commit to keep the history cleaner.

Comment (chmorgan) [reply PacketDotNet/DrdaDDMPacket.cs]: And this applies for all improvement changes in here that could go into the initial commit.

File: PacketDotNet/DrdaDDMPacket.cs [hunk 19]
```diff
@@ -159,9 +159,9 @@ public List<DrdaDDMParameter> Parameters
         }
 
         /// <summary>
-        /// 
+        /// Constructor
```
Comment (chmorgan) [inline PacketDotNet/DrdaDDMPacket.cs]: Please fold these and other improvement changes into the earlier commits when the files were introduced.

File: .vs/VSWorkspaceState.json [hunk 20]
```diff
@@ -1,7 +0,0 @@
-{
```
Comment (chmorgan) [inline .vs/VSWorkspaceState.json]: Please git fixup into earlier commits so the files were never added initially.

File: Test/PacketType/DrdaPacketTest.cs (hunk 13 above)
Comment (AndrewDi) [reply Test/PacketType/DrdaPacketTest.cs]: Thanks for your suggest, I have modify this method to do not copy array anymore.

File: Test/PacketType/DrdaPacketTest.cs (hunk 14 above)
Comment (AndrewDi) [reply Test/PacketType/DrdaPacketTest.cs]: There are lots of different drda packet, so I have to finish decode drda procotol first,than I will write more unit test case

File: Test/PacketType/DrdaPacketTest.cs (hunk 13 above)
Comment (AndrewDi) [reply Test/PacketType/DrdaPacketTest.cs]: This method just to test Ebcdic To Ascii String Convert, no more other usage, so I feel test here is fine, but if you suggest to put it in a separate test file, I can do it too.

**Comment by chmorgan** [comment]
@AndrewDi note that the AppVeyor build appears to have found an issue with the string test that you added.

**Comment by chmorgan** [comment]
@AndrewDi and looking at your PR you can git rebase and fixup all of those commits to be a single commit to implement DRDA protocol support I think.

**Comment by AndrewDi** [comment]
It seems like "Test.PacketType.LinktypeNullCaptureTest" have some error

**Comment by AndrewDi** [comment]
Do all the update and squash to one commit,please have a look
and remind AppVeyor build found some error about "Test.PacketType.LinktypeNullCaptureTest"

**Comment by chmorgan** [comment]
Hi @andrewdi. Please rebase against the current master. I had introduced an issue with that code that should be fixed on master now as of yesterday. Rebasing against master should fix that error.

**Comment by AndrewDi** [comment]
It seems like error still there

**Comment by AndrewDi** [comment]
I have open an issue, you can try to convert pcapng file to pcap format to fix the error

**Comment by chmorgan** [comment]
@AndrewDi ahh I'll have to take a look at that.

**Comment by AndrewDi** [comment]
Once you fix that error,I will rebase on the lastest commit again,thanks！

**Comment by chmorgan** [comment]
Hi Andrew.

Is the issue that you've introduced a pcapng file and it isn't able to be
parsed? If so, why not convert it to pcap format?

Chris

**Comment by AndrewDi** [comment]
Ok,I fix the linktype_null_capture packet format error,and All checks have passed now.

**Comment by chmorgan** [comment]
Thanks for the fix to the link type issue. I caught that here under appveyor, not sure how I didn't see that before, and its resolved.

**Comment by chmorgan** [comment]
I fixed a few things, running through appveyor and then if it passes I'll merge back to master.
Thank you again for the contribution, sorry for the delay in merging it in.

**Comment by chmorgan** [comment]
Merged to master and pushed out, thanks again!
